include/deal.II/multigrid/mg_transfer_matrix_free.h

```diff
@@ -369,6 +381,99 @@ class MGTransferBlockMatrixFree : public MGTransferBase<LinearAlgebra::distribut
 //------------------------ templated functions -------------------------
 #ifndef DOXYGEN
 
+
+template <int dim, typename Number>
+template <typename Number2, int spacedim>
+void
+MGTransferMatrixFree<dim,Number>::
+interpolate_to_mg(const DoFHandler<dim,spacedim> &dof_handler,
```

Comment:
@kronbichler i reworked your function to fit into `MGTransferMatrixFree`. It works fine in the absence of hanging nodes. I am in the process of extending it to cover this case, but something is wrong there... So far, I could not spot it, I would appreciate if you could give this function a quick look, maybe you spot some obvious mis-understanding on my side. 

The illustrative fine-level scalar-valued field  (from the test `itnerpolate_to_mg`) is 
![screen shot 2017-09-20 at 16 18 39](https://user-images.githubusercontent.com/8023934/30651186-4521234c-9e25-11e7-8811-b8b580adbea5.png)
and I see the problem at the bottom left cell on 0-th level (the problem could also be on other levels, it's just the test is written in such a way that it starts with 0-th level)
![screen shot 2017-09-20 at 16 56 47](https://user-images.githubusercontent.com/8023934/30651253-6e7ee6ac-9e25-11e7-82fb-67ad54739a57.png)

p.s. the idea of the test is project something trivial which can be represented exactly even on the coarsest level and then compare to evaluation at quadrature points.

Reply:
i did more debugging and it looks like on the problematic cell the value on the top-right node (which is shared with finer level) is not set:
```
dofs: 1 4 3 5
values: -12.2 -11.4 -11.8 0
```
i will dig further...

Reply:
@kronbichler I know why this does not work: the problem is there already on the finest level (=1) on the cell below:

<img width="752" alt="screen shot 2017-09-21 at 12 56 11" src="https://user-images.githubusercontent.com/8023934/30692567-5c2aa980-9ecc-11e7-96e0-de8fb66cd69d.png">

which has 
```
dofs: 0 1 2 3
values: 0 0 0 -10.4
```
that is because `copy_indices` in MGTransferMatrixFree for the finest level only have
```
copy_indices[1]	38	3
copy_indices[1]	40	7
copy_indices[1]	43	12
copy_indices[1]	44	13
copy_indices[1]	46	17
copy_indices[1]	47	18
copy_indices[1]	48	19
copy_indices[1]	50	22
copy_indices[1]	54	28
...
copy_indices[0]	1	1
copy_indices[0]	2	2
copy_indices[0]	3	3
copy_indices[0]	4	4
copy_indices[0]	5	5
```
and therefore fine-scale solution does not get transfered correctly on this cell during `copy_to_mg()`. 

This design decision probably had to do with the fact that one transfers residuals, but I guess it won't hurt to actually let those interface DoFs on the fine scale to actually by set-up correctly?

p.s. Combining this with the error on 0-th level, i see that the fine-scale solution at the corner between the levels (DoF=5) gets transfered to level 0.

Reply:
@masterleinad @tjhei ping as you both work with GMG. Quick summary of the last three detailed posts: [this line](https://github.com/dealii/dealii/blob/master/source/multigrid/mg_transfer_internal.cc#L117-L119) make life miserable when transferring fine-scale solution vector to MG vector.  By commenting it out things work as expected here but `copy_from_mg()` may not anymore...

Reply:
> This design decision probably had to do with the fact that one often transfers residuals, but I guess it won't hurt to actually let those interface DoFs to actually be transferred?

The whole point of the `copy_indices` is that it sets up a problem that later gets solved on the fine level. In the multigrid algorithm, we need to set zero Dirichlet boundary conditions on the interface between different levels. There is not really a away around imposing the correct condition for the fine-level problem for the multilevel algorithm, so I do not see how that could be relaxed without creating all sorts of problems.

So the interpolation really needs a completely independent code path at this point. I don't know whether we want to keep another variant of `copy_indices` or if we should just do it outside `MGTransfer`. I don't have a strong opinion regarding the interface, but we can definitely not change those parts without risking trouble in the multigrid algorithm.

Reply:
> So the interpolation really needs a completely independent code path at this point

given that we might at some point need to transfer solution vectors for multi-gird as well (see Fully Approximated Scheme), then I would say we shall keep things within `MGTransfer`. How about I add a flag to the `internal::fill_copy_indices()` do switch to another desired behaviour for solution vectors? Then I can store additional sets of indices and maybe build them first they are needed. Given how elaborated `fill_copy_indices` is, a completed independent code path does not sound too appealing to me.

Reply:
One more comment about this function: Can't this be in one of the base classes already? It does not use any of the `MatrixFree` specifics, so it would probably better fit somewhere else...